File: 739_Daily_Temperatures.js
var dailyTemperatures = function(temperatures) {
    let answer = [];
    for(let i = 0; i < temperatures.length; i++){
        if(i === temperatures.length - 1){
            answer.push(0);
            break;
        }
        let t = temperatures[i];
        let j = i + 1;
        while(j < temperatures.length){
            if(temperatures[j] > t){
                answer.push(j - i);
                break;
            }
            j++;
        }
        if(j === temperatures.length) answer.push(0);
    }
    return answer;
};

File: 739_Daily_Temperatures.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        stack<int> st;
        vector<int> ans(temperatures.size());
        int i = temperatures.size() - 1;
        while(i >= 0){
            if(st.empty()){
                st.push(i);
                ans[i] = 0;
            }
            else{
                while(!st.empty() && temperatures[st.top()] <= temperatures[i]){
                    st.pop();
                }
                if(!st.empty()){
                    ans[i] = st.top() - i;
                    st.push(i);
                } else {
                    ans[i] = 0;
                    st.push(i);
                }
            }
            i--;
        }
        return ans;
    }
};
